### PR TITLE
Added ignoring deprecated rules for --select=ALL

### PR DESCRIPTION
**Pull request by WindowGenerator**
## Summary

#10342: Added ignoring deprecated rules for --select=ALL

## Test Plan

Changed integration tests with --select=ALL option

### CODE REVIEW

File: crates/ruff_linter/src/rule_selector.rs [hunk 1]
```diff
@@ -189,8 +189,9 @@ impl RuleSelector {
     /// Return all matching rules, regardless of rule group filters like preview and deprecated.
     pub fn all_rules(&self) -> impl Iterator<Item = Rule> + '_ {
         match self {
-            RuleSelector::All => RuleSelectorIter::All(Rule::iter()),
-
+            RuleSelector::All => {
+                RuleSelectorIter::All(Rule::iter().filter(|rule| !rule.is_deprecated()))
+            }
```
Comment (zanieb) [inline crates/ruff_linter/src/rule_selector.rs:194]: Per the comment prefacing this function, we should not do this filtering here.

**Comment by zanieb** [comment]
@MichaReiser that's a good call... maybe. It'd be easy to hold off merging until v0.4.0 which seems reasonable.

**Comment by MichaReiser** [comment]
It might be fine, at least from our versioning policy. 

The change is reasonable to me but it might be worth documenting that all selects all not deprecated rules to make it clear to users that they might "loose" rules even when upgrading to a new patch version

File: crates/ruff_linter/src/rule_selector.rs (hunk 1 above)
Comment (zanieb) [reply crates/ruff_linter/src/rule_selector.rs:194]: Hey @WindowGenerator just want to check in again as we get closer to 0.4.0, did you want to work on this still?

Comment (WindowGenerator) [reply crates/ruff_linter/src/rule_selector.rs:194]: Hi @zanieb!
Yes, I will do it before the end of the weekend.

Comment (zanieb) [reply crates/ruff_linter/src/rule_selector.rs:194]: Not a rush! Thanks for the reply.

**Comment by Alexdelia** [comment]
hi, this changed was planned for version `v0.4.0`, but moved to `v0.5.0`

what is the reason behind this choice?

**Comment by zanieb** [comment]
Hey @Alexdelia, we ended up releasing the parser in 0.4.0 and focused on delivery of that. We routinely move things to later milestones when they aren't high priority for the given release.

File: crates/ruff_linter/src/rule_selector.rs [hunk 2]
```diff
@@ -226,7 +225,8 @@ impl RuleSelector {
             // is turned on
             || ((rule.is_preview() || rule.is_nursery()) && preview_enabled && (self.is_exact() || !preview_require_explicit))
             // Deprecated rules are excluded in preview mode unless explicitly selected
-            || (rule.is_deprecated() && (!preview_enabled || self.is_exact()))
+            // Ignore deprecated rules for All option
+            || (rule.is_deprecated() && (!preview_enabled || self.is_exact()) && !matches!(self, RuleSelector::All { .. }))
```
Comment (zanieb) [inline crates/ruff_linter/src/rule_selector.rs:229]: Can you combine this comment with the one above? e.g. 

> Deprecated rules are excluded from ALL in stable require exact selection in preview

Comment (WindowGenerator) [reply crates/ruff_linter/src/rule_selector.rs]: Did it!

**Comment by MichaReiser** [comment]
I tested that running ruff without preview no longer raises `MissingTypeSelf` when using `--select ALL`.  `ANN` keeps being raised when using `--select ANN`. 

To me this raises the question if deprecated rules should be removed from all non-exact selectors. It now feels kind of inconsistent that using `ALL` deselect deprecated rules but `ANN` doesn't

**Comment by MichaReiser** [comment]
I'll go ahead with merging this. It's heading in the right direction but we might need to figure out how we want to handle deprecation in combination with other group selectors.

**Pull request merged**